hab pkg bulkupload origin creation option

jeremymv2 opened the pull request:
Closes https://github.com/habitat-sh/habitat/issues/7101

## UX ⚙️ 
```
vagrant@chef-builder:~$ ./hab pkg bulkupload --url https://chef-builder.test --create-origins transfer/
» Preparing to upload artifacts to the 'unstable' channel on https://chef-builder.test
→ Using transfer/artifacts for artifacts and transfer/keys for signing keys.
→ Found 48 artifact(s) for upload.
☁ Discovering origin names from artifact metadata..
✓  chef
✓  core
✓  effortless
→ Found 3 origin(s) that may need to be created.
Ø Your Builder id will own any origin created. Be sure this is what you intend!
Ownership transfer is not yet implemented and will require SQL commands. [Yes/no/quit] 
...
```

Signed-off-by: Jeremy J. Miller <jm@chef.io>

apriofrost commented:
Some nitpicks for the UX in the terminal. The goal is to keep the confirmation message consistent with what's in the GUI. Would love to hear @kagarmoe 's thoughts on it as well!
```
vagrant@chef-builder:~$ ./hab pkg bulkupload --url https://chef-builder.test transfer/
» Preparing to upload artifacts to the 'unstable' channel on https://chef-builder.test
→ Using transfer/artifacts for artifacts and transfer/keys for signing keys.
→ Found 48 artifact(s) for upload.
☁ Discovering origin names from artifact metadata..
✓  chef
✓  core
✓  effortless
→ Found 3 missing origin(s).
Ø Origins are required for uploading the artifacts. The Builder account that created the origin is the owner. Ownership transfer is not yet supported and will require SQL commands (link to docs if we have it?). 
Create the following origins under your Builder account? [Yes/no/quit] 
chef
core
effortless
...
```

chefsalim commented on components/hab/src/main.rs:
```diff
@@ -901,9 +906,37 @@ fn sub_pkg_bulkupload(ui: &mut UI, m: &ArgMatches<'_>) -> Result<()> {
+                     commands."
+                               .to_string())?;
+            if !ask_create_origins(ui)? {
+                process::exit(1);
```
process exit seems overkill - can we not just return here with an Ok(())?

jeremymv2 replied:
Handled here https://github.com/habitat-sh/habitat/pull/7133/commits/5c46b2f97331cb937da872a6638e69252395d043

chefsalim commented on components/hab/src/main.rs:
```diff
@@ -901,9 +906,37 @@ fn sub_pkg_bulkupload(ui: &mut UI, m: &ArgMatches<'_>) -> Result<()> {
+            ui.status(Status::Custom(Glyph::CheckMark, "".to_string()), origin)?;
+        }
+        ui.status(Status::Found,
+                  format!("{} origin(s) that may need to be created.", origins.len()))?;
```
I think we should only give this warning if we find origins that are NOT present.  This seems like a generic warning being given all the time?

jeremymv2 replied:
Agreed that is less useful then what I had intended.
Because the Builder API [origins](https://github.com/habitat-sh/builder/blob/1626c3eec4389329df9044ae5a9075c0faba89e1/components/builder-api/doc/api.raml#L780) endpoint doesn't currently implement a `GET` or list functionality to enumerate the present origins, we just have to assume none of the origins exist and call `command::origin::create::start()` on each. 
Due to all of the above, I've just removed the message entirely in https://github.com/habitat-sh/habitat/pull/7133/commits/0c96916156ce8d0eeb4636319945c48c42d56bc6

jeremymv2 replied:
It seems that I am mistaken. The Builder API **does** implement a `GET` response on an origin name. 

**Note**: It is slightly more expensive to implement a `GET` followed by a `POST` as opposed to just blindly issuing a `POST` for the same end result. The existence check also adds some minor logic complexity as an additional cost.

The question then becomes: Is the additional existence check worthwhile to be able to provide the UX of only displaying the origins that need to be created?

markan replied:
I think it might be worthwhile to do the extra work, and change the flow a little bit.

Consider the case of a sync after the initial bootstrap; all the origins will be in place (most likely). I think the user would like to be warned if a new origin was needed, rather than have to mentally go through and say 'ok, core is ok, chef is ok, etc'.

If I understand the current behavior, if create_origins is false, and and an origin does not exist, we will go through the upload process until we try to upload a package to a non-existent origin, where we will then error. Given that the user's attention might have shifted to something else while waiting for uploads, that can create unnecessary delay and friction. 

Maybe we should instead always check the origins, and list missing origins. If create_origins is true, we go on and offer to create them, and if it's false we fail early. 

The main downside is the walk through the artifacts path and the 'GET' from builder will impose some extra work. I have some ideas for optimizations if that turns out to be expensive.

jeremymv2 replied:
@markan @apriofrost @chefsalim with this [commit](https://github.com/habitat-sh/habitat/pull/7133/commits/9f726c1fe1d6a3a53fa47c00d9f2ef5340de6c0a) there is only one CLI argument related to origin creation ("--auto-create-origins"). That is in alignment with @apriofrost's UX [diagram](https://github.com/habitat-sh/habitat/pull/7133#pullrequestreview-310218276) above. I believe that changes some of your reasoning above.

It also ends up meaning that origins are always auto created, the only difference is whether or not we prompt the user for confirmation.

Keeping in mind that `ommand::origin::create::start()` on each origin issues a `POST` but handles any existing origins gracefully:

```
Ω Creating origin chef.
… Skipping creation of origin chef. Origin already exists!
```

Due to the fact that in the current implementation origins are _always_ created and the creation method is effectively idempotent, is it worth it to implement a check?

jeremymv2 replied:
I think I'll go ahead and implement an existence check for the sake of completeness and elimination of any assumptions 😃

markan replied:
Thanks for the clarification of the behavior!

I still lean towards the existence check. I'm thinking about the sync case, where most of time the origins will exist. If we list those every time, even the ones that exist already, it will train the user to hit 'yes' (or set auto-create). If a new origin appears the chances are that reflex will have them say 'yes' because that's become the pattern. It's possible they will create an origin they didn't  intend to that way.

jeremymv2 replied:
I've implemented the client for the origin API check in https://github.com/habitat-sh/habitat/pull/7133/commits/bc818e341240f108e0210ce9c7c27bd80e522fed

chefsalim commented on components/hab/src/main.rs:
```diff
@@ -901,9 +906,37 @@ fn sub_pkg_bulkupload(ui: &mut UI, m: &ArgMatches<'_>) -> Result<()> {
+                  format!("{} origin(s) that may need to be created.", origins.len()))?;
+        if !origins.is_empty() && !accept {
+            ui.warn("Your Builder id will own any origin created. Be sure this is what you \
+                     intend!\nOwnership transfer is not yet implemented and will require SQL \
```
This message seems a bit heavy to me.  We should fast follow with the origin ownership transfer API/cli so we dont have to put in draconian warnings that will live around for a long time and be incorrect as soon as that API lands.

jeremymv2 replied:
This Builder issue exists for ownership transfer https://github.com/habitat-sh/builder/issues/36

jeremymv2 commented on components/hab/src/main.rs:
```diff
@@ -901,9 +906,37 @@ fn sub_pkg_bulkupload(ui: &mut UI, m: &ArgMatches<'_>) -> Result<()> {
+    if create_origins {
+        let mut origins: Vec<String> = Vec::new();
+        ui.status(Status::Discovering,
+                  "origin names from artifact metadata..".to_string())?;
```
Handled in https://github.com/habitat-sh/habitat/pull/7133/commits/3d92d7571c18c2d95e72072a930114c5caacc0dc

jeremymv2 commented on components/hab/src/main.rs:
```diff
@@ -901,9 +906,37 @@ fn sub_pkg_bulkupload(ui: &mut UI, m: &ArgMatches<'_>) -> Result<()> {
+            origins.push(ident.origin);
+        }
+        origins.sort();
+        origins.dedup();
```
Great callout!
Handled in https://github.com/habitat-sh/habitat/pull/7133/commits/73624f731d3319c8ed2cd6ef56932c99f0d7ac18

jeremymv2 commented:
> Some nitpicks for the UX in the terminal. The goal is to keep the confirmation message consistent with what's in the GUI. 

@apriofrost I've addressed your feedback wrt utilizing `--auto-create-origins` in this [commit](https://github.com/habitat-sh/habitat/pull/7133/commits/9f726c1fe1d6a3a53fa47c00d9f2ef5340de6c0a)

apriofrost commented on components/hab/src/cli.rs:
```diff
@@ -566,8 +566,7 @@ pub fn get(feature_flags: FeatureFlag) -> App<'static, 'static> {
                 (@arg AUTO_BUILD: --("auto-build") "Enable auto-build for all packages in this upload. Only applicable to SaaS Builder.")
-                (@arg CREATE_ORIGINS: --("create-origins") "Create any origins that do not exist in targeted Builder.")
-                (@arg ACCEPT: --("accept") "Do not prompt for confirmation of auto creation and ownership of any origins that do not exist.")
+                (@arg AUTO_CREATE_ORIGINS: --("auto-create-origins") "Do not prompt for confirmation before auto creating any origins that do not exist.")
```
I'd suggest using a description like this: "Skip the confirmation and automatically create origins that do not exist in the target Builder."
The two "do not"  in a sentence may make it a little bit harder to understand.

jeremymv2 replied:
Thanks, adjusted  in https://github.com/habitat-sh/habitat/pull/7133/commits/5f5cafefb4ab581f26d69e4f59770c099d5ae378

apriofrost replied:
Awesome! Thanks so much @jeremymv2 !

markan commented on components/hab/src/command/pkg/bulkupload.rs:
```diff
@@ -0,0 +1,125 @@
+            Err(err) => return Err(Error::from(err)),
+        }
+    }
+    if !auto_create_origins && !origins_to_create.is_empty() {
```
Tastes vary, but I found this a little bit confusing because of the reliance on the fall through case either being auto create or empty list. You have to remember the invariant when looking at the for loop. Maybe the following would be clearer.
```
if !origins_to_create.is_empty() {
   if !auto_create_origins  {
        ui.warn(String::from("Origins are required for uploading the artifacts. The Builder \
                              account that creates the origin is the owner."))?;
        if !ask_create_origins(ui)? {
            return Ok(());
        };
    };
    for origin_to_create in origins_to_create {
        command::origin::create::start(ui, &bldr_url, &token, &origin_to_create)?;
    }
}

chefsalim commented on components/hab/src/command/pkg/bulkupload.rs:
```diff
@@ -0,0 +1,126 @@
+    }
+    let mut origins_to_create: Vec<String> = Vec::new();
+    let api_client = Client::new(bldr_url, PRODUCT, VERSION, None)?;
+    for origin in origins {
```
Super tiny nit - add a whitespace blank line to break up the code sections a bit more (readability)

jeremymv2 replied:
Thank you: addressed in https://github.com/habitat-sh/habitat/pull/7133/commits/d13255c9d570503e6694cb0975e82350426e5314

jeremymv2 commented on components/hab/src/command/pkg/bulkupload.rs:
```diff
@@ -0,0 +1,126 @@
+    ui.status(Status::Found,
+              format!("{} artifact(s) for upload.", artifact_paths.len()))?;
+    ui.status(Status::Discovering,
+              String::from("origin names from local artifact cache.."))?;
```
Thanks, handled in https://github.com/habitat-sh/habitat/pull/7133/commits/d13255c9d570503e6694cb0975e82350426e5314

jeremymv2 commented on components/hab/src/command/pkg/bulkupload.rs:
```diff
@@ -0,0 +1,126 @@
+            Err(err) => return Err(Error::from(err)),
+        }
+    }
+    if !origins_to_create.is_empty() {
```
Thanks, I've handled in https://github.com/habitat-sh/habitat/pull/7133/commits/d13255c9d570503e6694cb0975e82350426e5314
